src/NorthwindWebsite/Presentation/Controllers/CategoriesController.cs

```diff
@@ -20,4 +26,64 @@ public async Task<IActionResult> Index()
 
         return View("Index", categories);
     }
+
+    [HttpGet]
+    public async Task<IActionResult> ImageUpload(int id)
+    {
+        var fileUploadModel = await _categoryService.GetFileUploadModel(id);
+
+        return View("Upload", fileUploadModel);
+    }
+
+    [HttpPost]
+    [ValidateAntiForgeryToken]
+    public async Task<IActionResult> ImageUpload(FileUploadDto fileUploadModel)
+    {
+        fileUploadModel.MaximumFileSize = _appSettings.FileUploadOptions.ImageMaxSize;
+
+        if (!ModelState.IsValid)
+        {
+            ModelState.AddModelError(string.Empty, "Please select a file.");
+            return View("Upload", fileUploadModel);
+        }
+
+        var permittedExtensions = _appSettings.FileUploadOptions.ImageFileFormats
+            .Split(", ").ToArray();
+
+        var uploadedFileExtension = fileUploadModel.FileUpload.ContentType.Split("/")[1];
+
+        if (!permittedExtensions.Contains(uploadedFileExtension))
+        {
+            ModelState.AddModelError(string.Empty,
+                String.Format("Wrong file format({0}). Please use .jpg, .jpeg, .bmp, .png", uploadedFileExtension));
+            return View("Upload", fileUploadModel);
+        }
+
+        if (fileUploadModel.FileUpload.Length > fileUploadModel.MaximumFileSize)
+        {
+            ModelState.AddModelError(string.Empty,
+                string.Format("The file is too large({0} kB). Maximum file size is {1} kB.",
+                fileUploadModel.FileUpload.Length / 1000, fileUploadModel.MaximumFileSize / 1000));
+            return View("Upload", fileUploadModel);
+        }
+
+        await _categoryService.UpdateCategoryWithPicture(fileUploadModel);
```

Comment:
So, let's make a little bit in a different way: controller should not validate available size of image to save, available extensions - it's related to your _Service_ layer. The only condition you can leave here is `ModelState.IsValid`. Let's move the rest to your business layer.